scene-lib/src/scenelib/annotations/el/TypePathEntry.java

```diff
@@ -0,0 +1,178 @@
+package scenelib.annotations.el;
+
+import org.checkerframework.checker.interning.qual.InternedDistinct;
+import org.objectweb.asm.TypePath;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A TypePathEntry is a way to get from one node in a {@link TypePath} to another.
+ * One can treat these as edges in a graph.
+ *
+ * A TypePathEntry corresponds
+ * to a step in an ASM {@link TypePath}.
+ *
+ * {@code List<TypePathEntry>} corresponds to an ASM {@link TypePath}.
+ * {@code List<TypePathEntry>} also corresponds to the javac class
+ * {@code com.sun.tools.javac.code.TypeAnnotationPosition}.
+ *
+ * {@code TypePathEntry} is immutable.
+ */
+public class TypePathEntry {
+  /**
+   * The kind of TypePathEntry; that is, how to get from the previous node in a TypePath to this one.
+   * One of TypePath.ARRAY_ELEMENT, TypePath.INNER_TYPE, TypePath.WILDCARD_BOUND, TypePath.TYPE_ARGUMENT.
+   *
+   * This corresponds to javac class {@code com.sun.tools.javac.code.TypeAnnotationPosition.TypePathEntry}.
+   */
+  public final int step;
+  /**
+   * If this represents a type argument (that is, step == TYPE_ARGUMENT), then the index for the type argument.
+   * Otherwise, 0.
+   */
+  public final int argument;
+
+  /** The canonical ARRAY_ELEMENT TypePathEntry for building TypePaths. */
+  public static final @InternedDistinct TypePathEntry ARRAY_ELEMENT = new TypePathEntry(TypePath.ARRAY_ELEMENT, 0);
+  /** The canonical INNER_TYPE TypePathEntry for building TypePaths. */
+  public static final @InternedDistinct TypePathEntry INNER_TYPE = new TypePathEntry(TypePath.INNER_TYPE, 0);
+  /** The canonical WILDCARD_BOUND TypePathEntry for building TypePaths. */
+  public static final @InternedDistinct TypePathEntry WILDCARD_BOUND = new TypePathEntry(TypePath.WILDCARD_BOUND, 0);
+
+  /**
+   * Construct a new TypePathEntry.
+   *
+   * @param step the type of the TypePathEntry
+   * @param argument index of the type argument or 0
+   */
+  public TypePathEntry(int step, int argument) {
```

Comment:
This is perhaps a small matter, I think the TypePathEntry constructor should be private.  Please use a factory method that returns one of `ARRAY_ELEMENT`, `INNER_TYPE`, or `WILDCARD_BOUND` when appropriate to avoid the overhead of constructing new objects.  When `step == TypePath.TYPE_ARGUMENT`, it's OK to construct a new object.